examples/optimize_evoked.py

```diff
@@ -0,0 +1,433 @@
+            consolidated_chunks[-1]['opt_end'] = max(consolidated_chunks[-1]['opt_end'], input_dict['opt_end'])
+            # average the weights
+            consolidated_chunks[-1]['weights'] = (consolidated_chunks[-1]['weights'] + input_dict['weights'])/2
+        else:
+            # new chunk
+            consolidated_chunks.append(input_dict)
+
+    # add one last chunk to the end
+    if len(consolidated_chunks) > 1:
+        last_chunk = create_last_chunk(consolidated_chunks)
+        consolidated_chunks.append(last_chunk)
+
+    return consolidated_chunks
+
+def optrun (new_params, grad=0):
+    """ This is the function to run a simulation
+    Parameters
+    ----------
+    new_params: Array
+        List or numpy array with the parameters chosen by
+        optimization engine. Order is consistent with
+        opt_params['ranges']
+    grad: float | None
+        Gradient if provided by optimization algorithm
+
+    Returns
+    -------
+    avg_rmse: float
+        Weighted RMSE between data in dpl and exp_dpl
+    """
+    global opt_params, params, exp_dpl, best_dpl, cur_step, optiter, stepminopterr
+    print("Optimization step %d, iteration %d"%(cur_step+1, optiter+1))
+
+    for param_name, test_value in zip(opt_params['ranges'].keys(), new_params): # set parameters
+        if test_value >= opt_params['ranges'][param_name]['minval'] and test_value <= opt_params['ranges'][param_name]['maxval']:
+            params[param_name] = test_value
+        else:
+            # This test is not strictly necessary with COBYLA, but in case the algorithm
+            # is changed at some point in the future
+            return 1e9 # invalid param value -> large error
+
+    # run the simulation, but stop early if possible
+    net = Network(params)
+    net.params['tstop'] = opt_params['opt_end']
+    dpls = simulate_dipole(net, n_jobs=1, n_trials=1)
+    avg_dpl = average_dipoles(dpls)
+    avg_rmse = rmse(avg_dpl, exp_dpl,
+                    tstart=opt_params['opt_start'],
+                    tstop=opt_params['opt_end'],
+                    weights=opt_params['weights'])
+
+    if avg_rmse < stepminopterr:
+        best = "[best] "
+        stepminopterr = avg_rmse
+        best_dpl = avg_dpl
+    else:
+        best = ""
+
+    print("%sweighted RMSE: %.2f over range [%3.3f-%3.3f] ms" % \
+        (best, avg_rmse, opt_params['opt_start'], opt_params['opt_end']))
+
+    optiter += 1
+
+    return avg_rmse # nlopt expects error
+
+def run_optimization(seed=0):
+    """ Start the nlopt optimization
+    Parameters
+    ----------
+    seed: int | None
+        Seed for RNG to make optimization results reproducible
+
+    Returns
+    -------
+    opt_results: List
+        List with the values of the best parameters found by
+        optimization engine. Order is consistent with
+        opt_params['ranges']
+    """
+    import nlopt
+    global opt_params
+
+    nlopt.srand(seed)
+    algorithm = nlopt.LN_COBYLA
```

Comment:
could we use scipy instead: https://docs.scipy.org/doc/scipy/reference/optimize.minimize-cobyla.html ?

Does nlopt give something that scipy does not?

Reply:
Oh, interesting! Feel free to change out for scipy. NLopt implemented PRAXIS for comparison and had better user documentation (scipy docs are complete but somewhat pedantic). Now that I've settled on using COBYLA, scipy seems preferable. I might make the change when the HNN optimization code gets rewritten.

Reply:
I looked into the change. Would take a bit more time to define constraint functions. Also not clear if the results are reproducible like nlopt (specifying seed).

Reply:
You can set the seed by providing `x0`